File: 2017/Case.FamilySubcategories/Case.FamilySubcategories/UI/form_Main.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Windows.Forms;
using Autodesk.Revit.UI;
using Microsoft.VisualBasic.ApplicationServices;
using clsSettings = Case.FamilySubcategories.Data.clsSettings;

namespace Case.FamilySubcategories.UI
{
  public partial class form_Main : Form
  {
    private readonly clsSettings _s;

    /// <summary>
    ///   Constructor
    /// </summary>
    /// <param name="s"></param>
    public form_Main(clsSettings s)
    {
      InitializeComponent();

      // Widen Scope
      _s = s;
    }

    #region Private Members

    private List<string> _fList = new List<string>(); //Family Path List 
    private string FolderPath { get; set; } //The folder
    
    /// <summary>
    /// Just the Text Headers
    /// </summary>
    private string HeaderTabs
    {
      get
      {
        return "Category\tFamily\tSource\tProjection Lineweight\tCut Lineweight\tLine Color\tMaterial\tSub-Categories";
      }
    }

    /// <summary>
    ///   Get the Data
    /// </summary>
    private void ProcessData()
    {
      try
      {
        if (_fList == null) return;
        if (saveFileDialog1.ShowDialog() != DialogResult.OK) return;
        if (saveFileDialog1.FileName == null) return;

        using (var sw = new StreamWriter(saveFileDialog1.FileName, false))
        {
          bool isFirstRow = true;

          for (int i = 0; i < _fList.Count; i++)
          {
            if (isFirstRow)
            {
              sw.WriteLine(HeaderTabs);
              isFirstRow = false;
            }
            string f = _fList[i];
            label1.Text = string.Format("Processing: {0}", Path.GetFileName(f));
            label1.Update();
            sw.WriteLine(_s.ProcessFamily(f));
            progressBar1.Increment(1);
          }

          sw.Close();
        }
      }
      catch (Exception ex)
      {
        Debug.WriteLine(ex.Message);
        using (var x = new TaskDialog("Error"))
        {
          x.TitleAutoPrefix = false;
          x.MainContent = "This blew up badly. Im not even sure what happened. Contact CASE.";
          x.MainInstruction = "I died, Jim.";
          x.Show();
        }
      }
      Close();
    }

    /// <summary>
    ///   Checks if there are any SubFolders
    /// </summary>
    /// <param name="path">The parent folder</param>
    /// <returns>bool if folders are present</returns>
    private bool HasSubfolder(string path)
    {
      var directory = new DirectoryInfo(path);
      DirectoryInfo[] subdirs = directory.GetDirectories();
      return (subdirs.Length != 0);
    }

    /// <summary>
    ///   Just a method for error processing
    /// </summary>
    /// <param name="path"></param>
    private void PathDoesNotExist(string path)
    {
      buttonOk.Enabled = false;

      if (string.IsNullOrEmpty(path)) path = "Empty Value";

      using (var x = new TaskDialog("Error"))
      {
        x.TitleAutoPrefix = false;
        x.MainContent = "Please browse to a valid path. \nInvalid Path: " + path;
        x.MainInstruction = "Invalid";
        x.Show();
      }
    }

    /// <summary>
    ///   Checks that the Directory is a Valid Path
    /// </summary>
    /// <param name="p">The Path</param>
    /// <returns>bool for Valid</returns>
    private bool TestPathInput(string p)
    {
      try
      {
        if (!Directory.Exists(p))
        {
          PathDoesNotExist(p);
          return false;
        }
        FolderPath = Path.GetFullPath(p);

        CheckSubFolders.Enabled = HasSubfolder(p);

        buttonOk.Enabled = true;

        return true;
      }
      catch (Exception ex)
      {
        Debug.WriteLine(ex.Message);
        PathDoesNotExist(p);
        return false;
      }
    }

    /// <summary>
    /// Give the user one last chance to get of it
    /// </summary>
    /// <returns>Go Forth?</returns>
    private bool LastChanceExit()
    {
      using (var x = new TaskDialog("Warning"))
      {
        x.TitleAutoPrefix = false;
        x.MainIcon = TaskDialogIcon.TaskDialogIconWarning;

        if (_fList.Any())
        {
          x.MainContent = "You are procesing " + _fList.Count +
                          " Families.\nThis will take approx. " +
                          Math.Round(TimeSpan.FromSeconds(_fList.Count*.5).TotalMinutes) +
                          " Minutes.\n" +
                          "This is your last chance to cancel this process.";
          x.MainInstruction = "Long wait time ahead.";
          x.CommonButtons = TaskDialogCommonButtons.Ok | TaskDialogCommonButtons.Cancel;
        }
        else
        {
          x.MainContent = "No Families Found in the selected directory";
          x.MainInstruction = "No Families Found";
          x.CommonButtons = TaskDialogCommonButtons.Cancel;
        }
        TaskDialogResult tdr = x.Show();
        return tdr == TaskDialogResult.Ok;
      }
    }

    #endregion

    #region Form Controls & Events

    /// <summary>
    ///   OK
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void buttonOk_Click(object sender, EventArgs e)
    {
      try
      {
        // Assembly Data for Tracking
        var m_a = new AssemblyInfo(Assembly.GetExecutingAssembly());

        // Usage Tracking        
////         var m_usage = new clsUsageFree("spicey", m_a);
      }
      catch (Exception exception)
      {
        Debug.WriteLine(exception.Message);
      }

      // Progress Bar Shuffle
      buttonOk.Hide();
      buttonCancel.Hide();
      ButtonHelp.Hide();

      _fList = _s.FamiliesToProcess(FolderPath, CheckSubFolders.Checked);

      progressBar1.Show();
      progressBar1.Minimum = 0;
      progressBar1.Value = 0;
      progressBar1.Maximum = _fList.Count;
      groupBox2.Hide();
      label1.Show();
      FormBorderStyle = FormBorderStyle.None;
      SizeGripStyle = SizeGripStyle.Hide;
      MinimumSize = new Size(400, 80);
      Size = new Size(400, 80);

      if (LastChanceExit())
      {
        ProcessData();
      }

      Close();
    }

    /// <summary>
    ///   Close
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void buttonCancel_Click(object sender, EventArgs e)
    {
      Close();
    }

    /// <summary>
    /// Goto the CaseApps page when someone needs help
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void HelpButton_Click(object sender, EventArgs e)
    {
      Process.Start("http://apps.case-inc.com/content/free-family-subcategories");
    }

    /// <summary>
    /// The Event when someone Clicks Browse
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void ButtonFolderBrowse_Click(object sender, EventArgs e)
    {
      if (FolderBrowser1.ShowDialog() != DialogResult.OK) return;
      TestPathInput(FolderBrowser1.SelectedPath);
      TextboxFamilyPath.Text = FolderPath;
    }

    /// <summary>
    /// The Evenet when someone changes the Text box 
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void TextboxFamilyPath_TextChanged(object sender, EventArgs e)
    {
      if (String.IsNullOrEmpty(TextboxFamilyPath.Text)) return;
      if (TestPathInput(TextboxFamilyPath.Text)) return;
      TextboxFamilyPath.Focus();
    }

    /// <summary>
    /// What happens when the form loads 
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void form_Main_Load(object sender, EventArgs e)
    {
      progressBar1.Hide();
      label1.Hide();
      buttonOk.Enabled = false;
    }

    #endregion

  }
}

File: 2018/Case.ChangeReplaceFamTypeNames/Case.ChangeReplaceFamTypeNames/UI/form_Main.vb
Imports Autodesk.Revit.DB
Imports Autodesk.Revit.UI
Imports System.Diagnostics
Imports [Case].ChangeReplaceFamTypeNames.API
Imports [Case].ChangeReplaceFamTypeNames.Data

Public Class form_Main

  Private _s As clsSettings

  ''' <summary>
  ''' Constructor
  ''' </summary>
  ''' <param name="s"></param>
  ''' <remarks></remarks>
  Public Sub New(s As clsSettings)
    InitializeComponent()

    ' Widen Scope
    _s = s
    AllowOK()
    ProgressBar1.Hide()
    Text = "Change & Replace Family/Type Names v" & _s.Version

  End Sub

  ''' <summary>
  ''' Minimum Text Requirements
  ''' </summary>
  ''' <remarks></remarks>
  Private Sub AllowOk()

    If Not (String.IsNullOrEmpty(TextBoxFind.Text) And String.IsNullOrEmpty(TextBoxReplace.Text)) Then
      If Not (CheckBoxFamilies.Checked = False And Me.CheckBoxTypes.Checked = False) Then
        ButtonOk.Enabled = True
      Else
        ButtonOk.Enabled = False
      End If
    Else
      ButtonOk.Enabled = False
    End If

  End Sub

  ''' <summary>
  ''' Rename the Items
  ''' </summary>
  ''' <remarks></remarks>
  Private Sub RenameStuff()

    ' Get the Families
    _s.GetFams()

    ' Only Rename Families Once
    Dim m_famFixed As New SortedDictionary(Of Integer, Integer)

    ' Progress Bar
    With Me.ProgressBar1
      .Value = 0
      .Minimum = 0
      .Maximum = _s.FamilyElements.Count
      .Show()
    End With

    ButtonOk.Hide()
    ButtonCancel.Hide()

    ' User Reporting
    Dim iFail As Integer = 0
    Dim iSuc As Integer = 0

    ' Process each Item
    For Each x In _s.FamilyElements

      ' Step the Progress Bar
      Try
        ProgressBar1.Increment(1)
      Catch
      End Try

      ' Renamed
      Dim m_renamed As Integer = 0

      ' Start a New Transaction
      Using t As New Transaction(_s.Doc, "CASE Renamer")
        If t.Start Then

          Try

            ' Check for Family Names
            If Me.CheckBoxFamilies.Checked = True Then

              ' Have we Processed the Family?
              If Not m_famFixed.ContainsKey(x.Family.Id.IntegerValue) Then

                ' Add it
                m_famFixed.Add(x.Family.Id.IntegerValue, x.Family.Id.IntegerValue)

                ' Check It
                If Not String.IsNullOrEmpty(TextBoxFind.Text) Then
                  If x.Family.Name.Contains(TextBoxFind.Text) Then
                    ' Rename This Part
                    Dim m_newName As String = Replace(x.Family.Name, Me.TextBoxFind.Text, Me.TextBoxReplace.Text, , , CompareMethod.Text)
                    x.Family.Name = m_newName
                    m_renamed += 1
                  End If
                End If

              End If

            End If

            ' Check for Type Names
            If Me.CheckBoxFamilies.Checked = True Then
              If Not String.IsNullOrEmpty(TextBoxFind.Text) Then
                If x.Name.Contains(TextBoxFind.Text) Then
                  ' Rename This Part
                  Dim m_newName As String = Replace(x.Name, Me.TextBoxFind.Text, Me.TextBoxReplace.Text, , , CompareMethod.Text)
                  x.Name = m_newName
                  m_renamed += 1
                End If
              End If
            End If

            ' Commit
            t.Commit()

            ' Count only if changes
            If m_renamed > 0 Then iSuc += m_renamed

          Catch
            iFail += 1
          End Try

        End If
      End Using

    Next

    ' Report to user
    Using td As New TaskDialog("Here's What Heppened:")
      td.MainInstruction = "Family and/or Type Renaming"
      td.MainContent = iSuc.ToString & " Updated Succesfully" & vbCr
      td.MainContent += iFail.ToString & " Items Failed to Rename"
      td.Show()
    End Using

  End Sub

#Region "Form Controls & Events"

  ''' <summary>
  ''' Help Documentation
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub ButtonHelp_Click(sender As System.Object, e As System.EventArgs) Handles ButtonHelp.Click
    Process.Start("http://apps.case-inc.com/content/free-change-replace-family-andor-type-names")
  End Sub

  ''' <summary>
  ''' Commit
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub ButtonOk_Click(sender As System.Object, e As System.EventArgs) Handles ButtonOk.Click

    ' Commit
    RenameStuff()

    ' Close
    Close()

  End Sub

  ''' <summary>
  ''' Close
  ''' </summary>
  ''' <param name="sender"></param>
  ''' <param name="e"></param>
  ''' <remarks></remarks>
  Private Sub ButtonCancel_Click(sender As System.Object, e As System.EventArgs) Handles ButtonCancel.Click

    ' Close
    Close()

  End Sub

  Private Sub TextBoxFind_TextChanged(sender As System.Object, e As System.EventArgs) Handles TextBoxFind.TextChanged
    AllowOK()
  End Sub

  Private Sub TextBoxReplace_TextChanged(sender As System.Object, e As System.EventArgs) Handles TextBoxReplace.TextChanged
    AllowOK()
  End Sub

  Private Sub CheckBoxFamilies_CheckedChanged(sender As System.Object, e As System.EventArgs) Handles CheckBoxFamilies.CheckedChanged
    AllowOK()
  End Sub

  Private Sub CheckBoxTypes_CheckedChanged(sender As System.Object, e As System.EventArgs) Handles CheckBoxTypes.CheckedChanged
    AllowOK()
  End Sub

  Private Sub PictureBox1_Click(sender As System.Object, e As System.EventArgs) Handles PictureBox1.Click
    Try
      Process.Start("http://apps.case-inc.com/")
    Catch
    End Try
  End Sub

#End Region

End Class